### PR TITLE
Xcframework prebuilt binaries

### PR DESCRIPTION
**Pull request by igstewart3**
This PR adds the ability to retrieve XCFramework prebuilt binaries. It installs them into the Carthage/Build folder and skips any Xcode.swift steps that involve checking the architecture.

It could probably use a wee bit of guidance from someone more familiar with the project to ensure I haven't missed something important.

I have tested this on a local project and was successfully able to install XCFramework binaries, as well as regular Framework binaries from a previous release.

Note: I have not been able to run the unit tests yet due to some Swift compile errors I haven't been able to resolve.

### CODE REVIEW

**Comment by sebskuse** [comment]
I've tested this against the latest https://github.com/aws-amplify/aws-sdk-ios, which provides prebuilt XCFramework binaries and this works as expected!

Using Carthage 0.37.0 we see:

```
*** Fetching aws-sdk-ios
*** Checking out aws-sdk-ios at "2.22.2"
*** No cache found for aws-sdk-ios, building with all downstream dependencies
*** xcodebuild output can be found in /var/folders/c0/068z__x16p15190541prsvkm0000gp/T/carthage-xcodebuild.RnWKmY.log
*** Downloading aws-sdk-ios.framework binary at "AWS SDK for iOS 2.22.2"
***  Skipped installing aws-sdk-ios.framework binary due to the error:
	"Invalid archive - Found multiple frameworks with the same unarchiving destination:
* 	file:///var/folders/c0/068z__x16p15190541prsvkm0000gp/T/carthage-archive.IXODoQ/AWSLogs.xcframework/ios-arm64_i386_x86_64-simulator/AWSLogs.framework/
	file:///var/folders/c0/068z__x16p15190541prsvkm0000gp/T/carthage-archive.IXODoQ/AWSLogs.xcframework/ios-arm64_armv7/AWSLogs.framework/ 
		to:
	file:///Path/To/Carthage/Build/iOS/AWSLogs.framework/

... repeat for every framework

Falling back to building from the source
```

Using this branch we see:

```
*** Fetching aws-sdk-ios
*** Checking out aws-sdk-ios at "2.22.2"
*** No cache found for aws-sdk-ios, building with all downstream dependencies
*** xcodebuild output can be found in /var/folders/c0/068z__x16p15190541prsvkm0000gp/T/carthage-xcodebuild.ll5AoF.log
*** Downloading aws-sdk-ios.framework binary at "AWS SDK for iOS 2.22.2"
```

And the prebuilt frameworks are in the expected folder.

**Comment by dmiluski** [comment]
Hi @igstewart3 ,

I just ran into this today with a similar hiccup with AWS (wowza **HUGE** install times). Thanks for kicking off the effort.

I'm not a repo owner/maintainer, but wanted to give some notes as I just started dabbling with this too:

Reading through I noticed this used string literals and bool logic to concatenate behaviors. Skimming the implementation, I noticed:

```
/// Describes the type of product built by an Xcode target.
public enum ProductType: String {
	/// A framework bundle.
	case framework = "com.apple.product-type.framework"

	/// A static library.
	case staticLibrary = "com.apple.product-type.library.static"

	/// A unit test bundle.
	case testBundle = "com.apple.product-type.bundle.unit-test"
```

And was wondering if perhaps xcframework should be included as a sibling here? Allowing for the use of this definition for path? Allowing for a model to drive potential differentiation rather than bool checks?

```
/// Framework Suffix
public enum FrameworkSuffix: String {

    /// Dynamic Framework
    case dynamic = "framework"

    /// XCFramework
    case xcode = "xcframework"
}
```

**Comment by igstewart3** [comment]
> Hi @igstewart3 ,
> 
> I just ran into this today with a similar hiccup with AWS (wowza **HUGE** install times). Thanks for kicking off the effort.
> 
> I'm not a repo owner/maintainer, but wanted to give some notes as I just started dabbling with this too:
> 
> Reading through I noticed this used string literals and bool logic to concatenate framework download behaviors. Skimming the implementation, I noticed:
> 
> ```
> /// Describes the type of product built by an Xcode target.
> public enum ProductType: String {
> 	/// A framework bundle.
> 	case framework = "com.apple.product-type.framework"
> 
> 	/// A static library.
> 	case staticLibrary = "com.apple.product-type.library.static"
> 
> 	/// A unit test bundle.
> 	case testBundle = "com.apple.product-type.bundle.unit-test"
> ```
> 
> And was wondering if perhaps xcframework should be included as a sibling here? Allowing for the use of this definition for path? Allowing for a model to drive potential differentiation rather than bool checks?
> 
> ```
> /// Framework Suffix
> public enum FrameworkSuffix: String {
> 
>     /// Dynamic Framework
>     case dynamic = "framework"
> 
>     /// XCFramework
>     case xcode = "xcframework"
> }
> ```

Good shout, I'll take a look at tidying that up 👍

**Comment by finik** [comment]
I can confirm as well, the fix in this branch works, please merge and release this ASAP, this blocks some frameworks from being properly distributed. Thanks

File: Source/CarthageKit/Project.swift [hunk 1]
```diff
@@ -1447,11 +1457,13 @@ func platformForFramework(_ frameworkURL: URL) -> SignalProducer<SDK, CarthageEr
 /// Sends the URL to each framework bundle found in the given directory.
 internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL, CarthageError> {
 	return filesInDirectory(directoryURL, kUTTypeFramework as String)
+        .concat(filesInDirectory(directoryURL, "com.apple.xcframework"))
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Project.swift]: Done 👍

File: Source/CarthageKit/Xcode.swift [hunk 2]
```diff
@@ -1413,6 +1419,9 @@ public func nonDestructivelyStripArchitectures(_ frameworkURL: URL, _ architectu
 
 /// Strips the given architectures from a framework.
 private func stripArchitectures(_ packageURL: URL, _ architectures: Set<String>) -> SignalProducer<(), CarthageError> {
+    guard isNotXCFramework(packageURL) else {
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Xcode.swift]: Done 👍

File: Source/CarthageKit/Xcode.swift [hunk 3]
```diff
@@ -1431,6 +1440,9 @@ private func stripArchitectures(_ packageURL: URL, _ architectures: Set<String>)
 
 // Returns a signal of all architectures present in a given package.
 public func architecturesInPackage(_ packageURL: URL, xcrunQuery: [String] = ["lipo", "-info"]) -> SignalProducer<[String], CarthageError> {
+    guard isNotXCFramework(packageURL) else {
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Xcode.swift]: Done for all changes made in the PR 👍

**Review by elliottwilliams — CHANGES_REQUESTED** [review]
First off, thanks so much for your contribution! I think this makes a lot of sense and is obviously something we need to support.

Aside from the suggestions below, we need to handle `checkFrameworkCompatibility` correctly. Since it's possible to build an xcframework with `-allow-internal-distribution`, it's possible to have swiftmodules in the xcframework that aren't compatible with the machine they're downloaded to. This patch should fix it, LMK if there's anything I can clarify:
[carthage-3123-additions.patch.txt](https://github.com/Carthage/Carthage/files/6056150/carthage-3123-additions.patch.txt)

File: Source/CarthageKit/Project.swift [hunk 4]
```diff
@@ -721,6 +714,23 @@ public final class Project { // swiftlint:disable:this type_body_length
...
+    
+    /// Ensures binary framework has a valid extension and returns url in build folder
+	private func getBinaryFrameworkURL(url: URL) -> SignalProducer<URL, CarthageError> {
+		return SignalProducer(result: FrameworkSuffix.from(string: url.pathExtension))
+			.flatMap(.merge) { frameworkSuffix -> SignalProducer<URL, CarthageError> in
+				switch frameworkSuffix {
+				case .xcframework:
+					return SignalProducer<URL, CarthageError>(value: url)
+						.attemptMap { self.frameworkURLInCarthageBuildFolder(forSDK: SDK(name: "", simulatorHeuristic: ""),
+																			 frameworkNameAndExtension: $0.lastPathComponent) }
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Project.swift]: Done 👍

File: Source/CarthageKit/Project.swift [hunk 5]
```diff
@@ -1447,11 +1457,13 @@ func platformForFramework(_ frameworkURL: URL) -> SignalProducer<SDK, CarthageEr
...
 internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL, CarthageError> {
 	return filesInDirectory(directoryURL, kUTTypeFramework as String)
+		.concat(filesInDirectory(directoryURL, "com.apple.xcframework"))
 		.filter { !$0.pathComponents.contains("__MACOSX") }
 		.filter { url in
 			// Skip nested frameworks
 			let frameworksInURL = url.pathComponents.filter { pathComponent in
-				return (pathComponent as NSString).pathExtension == "framework"
+				let pathExtension = (pathComponent as NSString).pathExtension
+				return FrameworkSuffix.from(string: pathExtension).error == nil
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Project.swift]: Done 👍

File: Source/CarthageKit/Project.swift [hunk 6]
```diff
@@ -721,6 +714,23 @@ public final class Project { // swiftlint:disable:this type_body_length
 					.then(SignalProducer<URL, CarthageError>(value: directoryURL))
 			}
 	}
+    
+    /// Ensures binary framework has a valid extension and returns url in build folder
+	private func getBinaryFrameworkURL(url: URL) -> SignalProducer<URL, CarthageError> {
+		return SignalProducer(result: FrameworkSuffix.from(string: url.pathExtension))
+			.flatMap(.merge) { frameworkSuffix -> SignalProducer<URL, CarthageError> in
+				switch frameworkSuffix {
+				case .xcframework:
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Project.swift]: Done 👍

File: Source/CarthageKit/Xcode.swift [hunk 7]
```diff
@@ -1649,3 +1667,7 @@ private func codesign(_ frameworkURL: URL, _ expandedIdentity: String) -> Signal
 		.mapError(CarthageError.taskError)
 		.then(SignalProducer<(), CarthageError>.empty)
 }
+
+private func isNotXCFramework(_ frameworkUrl: URL) -> Bool {
+	return FrameworkSuffix.from(string: frameworkUrl.pathExtension).value != .xcframework
+}
```
[the remark this answers is not in the document]
Comment (igstewart3) [reply Source/CarthageKit/Xcode.swift]: Several of these methods are called when installing the prebuilt XCFrameworks and fail due to not being able to find the executableURL in the bundle in `binaryURL(packageUrl:)`. We could look into adding a custom case for XCFrameworks like is already in place for .dSYM and (and changing to return an array of URLs), but many of these processes (stripping architectures and debug symbols) don't necessarily make sense for an XCFramework anyway.

**Comment by igstewart3** [comment]
@elliottwilliams Thanks very much for the feedback! I've implemented most of it, including the patch you provided. I think only the XCFramework guard cases in `Xcode.swift` still need to be resolved. They definitely allow the prebuilt binary to be installed successfully but they do feel a little clumsy, keen to hear your thoughts on that.

**Comment by elliottwilliams** [comment]
> @elliottwilliams Thanks very much for the feedback! I've implemented most of it, including the patch you provided. I think only the XCFramework guard cases in `Xcode.swift` still need to be resolved. They definitely allow the prebuilt binary to be installed successfully but they do feel a little clumsy, keen to hear your thoughts on that.

@igstewart3 I think I got it working – we just need to avoid calling into the dSYM/bcsymbolmap logic higher up. This is similar to [what we do on the `build` side](https://github.com/Carthage/Carthage/blob/709b82a2db586f419b18fa21e53ea0cd4ab70745/Source/CarthageKit/Xcode.swift#L776-L784), where we conditionally avoid calling any debug symbols merging code when making an xcframework. See the commit I just pushed :)

I confirmed this works with a `github "aws-amplify/aws-sdk-ios"` Cartfile, but LMK if there's anything else you're noticing.

**Comment by igstewart3** [comment]
@elliottwilliams Nice! Good fix, that looks a lot tidier. Everything seems to be working in my tests as well 👍

**Review by elliottwilliams — APPROVED** [review]
Sounds good! Going to give the other maintainers a moment to review if they'd like.

**Review by ozgur — APPROVED** [review]

**Pull request merged**